### PR TITLE
List all lowering implementation in developer docs

### PR DESCRIPTION
**Pull request by sklam**
With the growing number of implemented lowering and specializing overloads that can be in different source files, it becomes harder to find the definition location of the lowering of a supported nopython function.  This patch adds a autogeneration script that runs at documentation generation to add a section in the developer doc with a full listing of all lowering implementation.  In addition to help finding definition location of lowering implementation, this will also:

* show implemented nopython features and the supported signatures
* show how the lowering is implemented via source listing
* help understand the specialization/generialization for potential refactoring.

### CODE REVIEW

**Comment by seibert** [comment]
When Siu first showed me this, it seemed kind of crazy, but it actually becomes oddly compelling to browse through and learn how to write new lowerings.  We could put it in another section besides the developer docs, but I think the idea is interesting.

File: docs/source/developer/listings.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Listings
+========
+
```
Comment (seibert) [inline docs/source/developer/listings.rst]: This section definitely needs a more descriptive title, and a paragraph explanation of what the reader is looking at.

**Comment by pitrou** [comment]
Is there an online demo of what this outputs? :-)

File: docs/source/conf.py [hunk 2]
```diff
@@ -283,3 +283,16 @@
...
+
+def _autogenerate():
+    from numba.scripts.generate_lower_listing import gen_lower_listing
+
+    basedir = os.path.dirname(__file__)
+    gen_lower_listing(os.path.join(basedir,
+                                   'developer/autogen_lower_listing.rst'))
+
+
+_autogenerate()
```
Comment (pitrou) [inline docs/source/conf.py]: Ideally there should be a way to only run _autogenerate() when required. I suppose this is adding some latency to the docs build time.

File: numba/scripts/generate_lower_listing.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                print()
+
+                impl = info['impl']
+                print('by', '``{0}``'.format(impl['name']),
+                      '`{0}`'.format(impl['filename']),
+                      'lines {0}-{1}'.format(*impl['lines']))
+                print()
+                print(".. code-block:: python")
+                print()
+                print(textwrap.indent(impl['source'], prefix=' ' * 4))
```
Comment (pitrou) [inline numba/scripts/generate_lower_listing.py]: Is this really including the source code of all functions?

File: docs/source/conf.py (hunk 2 above)
Comment (sklam) [inline docs/source/conf.py]: A few seconds

File: numba/scripts/generate_lower_listing.py (hunk 3 above)
Comment (sklam) [inline numba/scripts/generate_lower_listing.py]: Hee =) right

**Comment by pitrou** [comment]
Ok, so I ran the script. I agree the list of signatures for each implemented function is an interesting read, but I really don't like the inclusion of source code. It makes the page massive and difficult to navigate accross. I think it's enough to write the source files and line numbers beside each signature.

Also, you can make hypertext links to the source code on Github, for example "numba/targets/tupleobj.py lines 77-80" is at https://github.com/numba/numba/blob/master/numba/targets/tupleobj.py#L77-L80 (so the pattern appears to be `https://github.com/numba/numba/blob/master/{filepath}#L{firstlineno}-L{lastlineno}`)

File: docs/source/conf.py (hunk 2 above)
Comment (pitrou) [inline docs/source/conf.py]: Apparently it's 12 seconds here.

**Comment by sklam** [comment]
I like the idea of hyperlink to GH.  Should be possible to make it point exactly to the commit from which the doc is generated.

**Comment by seibert** [comment]
yeah, I have found the source links in the numpy docs to be invaluable.

**Comment by pitrou** [comment]
Also, hopefully not including the source code will reduce the processing time (although the clean way to solve this may be to write a Sphinx extension).

File: docs/source/conf.py (hunk 2 above)
Comment (pitrou) [inline docs/source/conf.py]: With the latest version (not including the source code anymore) it feels much faster (one second or two). Thanks!

**Comment by pitrou** [comment]
I think this is good enough now, we can fine-tune later. @seibert, what do you think?

**Comment by seibert** [comment]
Yeah, with the switch to github links, I'm happy with this.

**Pull request merged**